Flaky test: Test/RetryStats

```
--- FAIL: Test (132.29s)
    --- FAIL: Test/RetryStats (0.07s)
        tlogger.go:116: INFO clientconn.go:175 [core] Channel Created  (t=+877.105µs)
        tlogger.go:116: INFO clientconn.go:252 [core] parsed scheme: ""  (t=+1.064006ms)
        tlogger.go:116: INFO clientconn.go:258 [core] scheme "" not registered, fallback to default scheme  (t=+1.171306ms)
        tlogger.go:116: INFO resolver_conn_wrapper.go:100 [core] ccResolverWrapper: sending update to cc: {[{127.0.0.1:33815  <nil> 0 <nil>}] <nil> <nil>}  (t=+1.239107ms)
        tlogger.go:116: INFO resolver_conn_wrapper.go:183 [core] Resolver state updated: {Addresses:[{Addr:127.0.0.1:33815 ServerName: Attributes:<nil> Type:0 Metadata:<nil>}] ServiceConfig:<nil> Attributes:<nil>} (resolver returned new addresses)  (t=+1.315407ms)
        tlogger.go:116: INFO clientconn.go:728 [core] ClientConn switching balancer to "pick_first"  (t=+1.384808ms)
        tlogger.go:116: INFO clientconn.go:748 [core] Channel switches to new LB policy "pick_first"  (t=+1.591409ms)
        tlogger.go:116: INFO clientconn.go:789 [core] Subchannel Created  (t=+1.79561ms)
        tlogger.go:116: INFO clientconn.go:789 [core] Subchannel(id:2669) created  (t=+1.88151ms)
        tlogger.go:116: INFO clientconn.go:1142 [core] Subchannel Connectivity change to CONNECTING  (t=+2.238412ms)
        tlogger.go:116: INFO clientconn.go:1253 [core] Subchannel picks a new address "127.0.0.1:33815" to connect  (t=+2.675915ms)
        tlogger.go:116: INFO picker_wrapper.go:161 [core] blockingPicker: the picked transport is not ready, loop back to repick  (t=+3.188118ms)
        tlogger.go:116: INFO clientconn.go:436 [core] Channel Connectivity change to CONNECTING  (t=+3.49202ms)
        tlogger.go:116: INFO clientconn.go:1142 [core] Subchannel Connectivity change to READY  (t=+4.058423ms)
        tlogger.go:116: INFO clientconn.go:436 [core] Channel Connectivity change to READY  (t=+4.476325ms)
        retry_test.go:604: Handler received frames:
            *stats.Begin - &{true 2021-09-22 23:53:07.696314257 +0000 UTC m=+294.826907066 true false false false}
            *stats.OutHeader - &{true  map[user-agent:[grpc-go/1.42.0-dev]] /grpc.testing.TestService/EmptyCall 127.0.0.1:33815 127.0.0.1:52078}
            *stats.OutPayload - &{true  [] 0 5 2021-09-22 23:53:07.698925672 +0000 UTC m=+294.829518381}
            *stats.End - &{true 2021-09-22 23:53:07.696314257 +0000 UTC m=+294.826907066 2021-09-22 23:53:07.699691276 +0000 UTC m=+294.830284085 map[] rpc error: code = Unavailable desc = stream terminated by RST_STREAM with error code: REFUSED_STREAM}
            *stats.Begin - &{true 2021-09-22 23:53:07.699734076 +0000 UTC m=+294.830326785 true false false true}
            *stats.OutHeader - &{true  map[user-agent:[grpc-go/1.42.0-dev]] /grpc.testing.TestService/EmptyCall 127.0.0.1:33815 127.0.0.1:52078}
            *stats.OutPayload - &{true <nil> [] 0 5 2021-09-22 23:53:07.700072178 +0000 UTC m=+294.830664887}
            *stats.InTrailer - &{true 73 map[content-type:[application/grpc] grpc-retry-pushback-ms:[10]]}
            *stats.End - &{true 2021-09-22 23:53:07.699734076 +0000 UTC m=+294.830326785 2021-09-22 23:53:07.700773782 +0000 UTC m=+294.831366591 map[content-type:[application/grpc] grpc-retry-pushback-ms:[10]] rpc error: code = Unavailable desc = unavailable retry}
            *stats.Begin - &{true 2021-09-22 23:53:07.711544443 +0000 UTC m=+294.842137152 true false false false}
            *stats.OutHeader - &{true  map[user-agent:[grpc-go/1.42.0-dev]] /grpc.testing.TestService/EmptyCall 127.0.0.1:33815 127.0.0.1:52078}
            *stats.InHeader - &{true 14  map[content-type:[application/grpc]]  <nil> <nil>}
            *stats.InTrailer - &{true 12 map[]}
            *stats.OutPayload - &{true <nil> [] 0 5 2021-09-22 23:53:07.712593649 +0000 UTC m=+294.843186358}
            *stats.InPayload - &{true  [] 0 5 2021-09-22 23:53:07.71275705 +0000 UTC m=+294.843349859}
            *stats.End - &{true 2021-09-22 23:53:07.711544443 +0000 UTC m=+294.842137152 2021-09-22 23:53:07.71280705 +0000 UTC m=+294.843399859 map[] <nil>}
            ---
            want:
            *stats.Begin - &{false 0001-01-01 00:00:00 +0000 UTC false false false false}
            *stats.OutHeader - &{false  map[] /grpc.testing.TestService/EmptyCall <nil> <nil>}
            *stats.OutPayload - &{false <nil> [] 0 5 0001-01-01 00:00:00 +0000 UTC}
            *stats.End - &{false 0001-01-01 00:00:00 +0000 UTC 0001-01-01 00:00:00 +0000 UTC map[] <nil>}
            *stats.Begin - &{false 0001-01-01 00:00:00 +0000 UTC false false false true}
            *stats.OutHeader - &{false  map[] /grpc.testing.TestService/EmptyCall <nil> <nil>}
            *stats.OutPayload - &{false <nil> [] 0 5 0001-01-01 00:00:00 +0000 UTC}
            *stats.InTrailer - &{false 0 map[content-type:[application/grpc] grpc-retry-pushback-ms:[10]]}
            *stats.End - &{false 0001-01-01 00:00:00 +0000 UTC 0001-01-01 00:00:00 +0000 UTC map[] <nil>}
            *stats.Begin - &{false 0001-01-01 00:00:00 +0000 UTC false false false false}
            *stats.OutHeader - &{false  map[] /grpc.testing.TestService/EmptyCall <nil> <nil>}
            *stats.OutPayload - &{false <nil> [] 0 5 0001-01-01 00:00:00 +0000 UTC}
            *stats.InHeader - &{false 0  map[]  <nil> <nil>}
            *stats.InPayload - &{false <nil> [] 0 5 0001-01-01 00:00:00 +0000 UTC}
            *stats.InTrailer - &{false 0 map[]}
            *stats.End - &{false 0001-01-01 00:00:00 +0000 UTC 0001-01-01 00:00:00 +0000 UTC map[] <nil>}
        retry_test.go:628: at position 11: got *stats.InHeader; want *stats.OutPayload
        tlogger.go:116: INFO clientconn.go:436 [core] Channel Connectivity change to SHUTDOWN  (t=+20.612116ms)
        tlogger.go:116: INFO clientconn.go:1142 [core] Subchannel Connectivity change to SHUTDOWN  (t=+20.968918ms)
        tlogger.go:116: INFO clientconn.go:1504 [core] Subchannel Deleted  (t=+21.18772ms)
        tlogger.go:116: INFO clientconn.go:1504 [core] Subchanel(id:2669) deleted  (t=+21.24462ms)
        tlogger.go:116: INFO clientconn.go:1098 [core] Channel Deleted  (t=+21.391721ms)
        tlogger.go:116: INFO types.go:68 [channelz] attempt to delete child with id 2669 from a parent (id=2668) that doesn't currently exist  (t=+22.813729ms)
        tlogger.go:116: INFO types.go:68 [channelz] attempt to delete child with id 2669 from a parent (id=2668) that doesn't currently exist  (t=+23.721134ms)
FAIL
FAIL	google.golang.org/grpc/test	307.110s
```

https://github.com/grpc/grpc-go/pull/4807/checks?check_run_id=3681386419

--- a/internal/transport/http2_client.go
+++ b/internal/transport/http2_client.go
@@ -1073,7 +1073,7 @@ func (t *http2Client) handleData(f *http2.DataFrame) {
 	}
 	// The server has closed the stream without sending trailers.  Record that
 	// the read direction is closed, and set the status appropriately.
-	if f.FrameHeader.Flags.Has(http2.FlagDataEndStream) {
+	if f.StreamEnded() {
 		t.closeStream(s, io.EOF, false, http2.ErrCodeNo, status.New(codes.Internal, "server closed the stream without sending trailers"), nil, true)
 	}
 }

--- a/internal/transport/http2_server.go
+++ b/internal/transport/http2_server.go
@@ -734,7 +734,7 @@ func (t *http2Server) handleData(f *http2.DataFrame) {
 			s.write(recvMsg{buffer: buffer})
 		}
 	}
-	if f.Header().Flags.Has(http2.FlagDataEndStream) {
+	if f.StreamEnded() {
 		// Received the end of stream from the client.
 		s.compareAndSwapState(streamActive, streamReadDone)
 		s.write(recvMsg{err: io.EOF})

--- a/test/end2end_test.go
+++ b/test/end2end_test.go
@@ -7352,8 +7352,11 @@ type httpServerResponse struct {
 }
 
 type httpServer struct {
-	refuseStream func(uint32) bool
-	responses    []httpServerResponse
+	// If waitForEndStream is set, wait for the client to send a frame with end
+	// stream in it before sending a response/refused stream.
+	waitForEndStream bool
+	refuseStream     func(uint32) bool
+	responses        []httpServerResponse
 }
 
 func (s *httpServer) writeHeader(framer *http2.Framer, sid uint32, headerFields []string, endStream bool) error {
@@ -7416,8 +7419,14 @@ func (s *httpServer) start(t *testing.T, lis net.Listener) {
 					}
 					return
 				}
-				if hframe, ok := frame.(*http2.HeadersFrame); ok {
+				sid = 0
+				if hframe, ok := frame.(*http2.HeadersFrame); ok && (!s.waitForEndStream || hframe.StreamEnded()) {
 					sid = hframe.Header().StreamID
+				}
+				if dframe, ok := frame.(*http2.DataFrame); ok && s.waitForEndStream && dframe.Header().Flags.Has(http2.FlagDataEndStream) {
+					sid = dframe.Header().StreamID
+				}
+				if sid != 0 {
 					if s.refuseStream == nil || !s.refuseStream(sid) {
 						break
 					}

--- a/test/end2end_test.go
+++ b/test/end2end_test.go
@@ -7422,12 +7422,16 @@ func (s *httpServer) start(t *testing.T, lis net.Listener) {
 				sid = 0
 				switch fr := frame.(type) {
 				case *http2.HeadersFrame:
+					// Respond after this if we are not waiting for an end
+					// stream or if this frame ends it.
 					if !s.waitForEndStream || fr.StreamEnded() {
 						sid = fr.Header().StreamID
 					}
 
 				case *http2.DataFrame:
-					if s.waitForEndStream && fr.Header().Flags.Has(http2.FlagDataEndStream) {
+					// Respond after this if we were waiting for an end stream
+					// and this frame ends it.
+					if s.waitForEndStream && fr.StreamEnded() {
 						sid = fr.Header().StreamID
 					}
 				}

--- a/test/end2end_test.go
+++ b/test/end2end_test.go
@@ -7420,11 +7420,16 @@ func (s *httpServer) start(t *testing.T, lis net.Listener) {
 					return
 				}
 				sid = 0
-				if hframe, ok := frame.(*http2.HeadersFrame); ok && (!s.waitForEndStream || hframe.StreamEnded()) {
-					sid = hframe.Header().StreamID
-				}
-				if dframe, ok := frame.(*http2.DataFrame); ok && s.waitForEndStream && dframe.Header().Flags.Has(http2.FlagDataEndStream) {
-					sid = dframe.Header().StreamID
+				switch fr := frame.(type) {
+				case *http2.HeadersFrame:
+					if !s.waitForEndStream || fr.StreamEnded() {
+						sid = fr.Header().StreamID
+					}
+
+				case *http2.DataFrame:
+					if s.waitForEndStream && fr.Header().Flags.Has(http2.FlagDataEndStream) {
+						sid = fr.Header().StreamID
+					}
 				}
 				if sid != 0 {
 					if s.refuseStream == nil || !s.refuseStream(sid) {

--- a/test/end2end_test.go
+++ b/test/end2end_test.go
@@ -7430,7 +7430,9 @@ func (s *httpServer) start(t *testing.T, lis net.Listener) {
 
 				case *http2.DataFrame:
 					// Respond after this if we were waiting for an end stream
-					// and this frame ends it.
+					// and this frame ends it.  (If we were not waiting for an
+					// end stream, this stream was already responded to when
+					// the headers were received.)
 					if s.waitForEndStream && fr.StreamEnded() {
 						sid = fr.Header().StreamID
 					}

--- a/test/retry_test.go
+++ b/test/retry_test.go
@@ -517,6 +517,7 @@ func (s) TestRetryStats(t *testing.T) {
 	}
 	defer lis.Close()
 	server := &httpServer{
+		waitForEndStream: true,
 		responses: []httpServerResponse{{
 			trailers: [][]string{{
 				":status", "200",
@@ -588,13 +589,6 @@ func (s) TestRetryStats(t *testing.T) {
 		&stats.End{},
 	}
 
-	// There is a race between noticing the RST_STREAM during the first RPC
-	// attempt and writing the payload.  If we detect that the client did not
-	// send the OutPayload, we remove it from want.
-	if _, ok := handler.s[2].(*stats.End); ok {
-		want = append(want[:2], want[3:]...)
-	}
-
 	toString := func(ss []stats.RPCStats) (ret []string) {
 		for _, s := range ss {
 			ret = append(ret, fmt.Sprintf("%T - %v", s, s))
@@ -612,8 +606,7 @@ func (s) TestRetryStats(t *testing.T) {
 	// There is a race between receiving the payload (triggered by the
 	// application / gRPC library) and receiving the trailer (triggered at the
 	// transport layer).  Adjust the received stats accordingly if necessary.
-	// Note: we measure from the end of the RPCStats due to the race above.
-	tIdx, pIdx := len(handler.s)-3, len(handler.s)-2
+	const tIdx, pIdx = 13, 14
 	_, okT := handler.s[tIdx].(*stats.InTrailer)
 	_, okP := handler.s[pIdx].(*stats.InPayload)
 	if okT && okP {
@@ -654,8 +647,8 @@ func (s) TestRetryStats(t *testing.T) {
 	}
 
 	// Validate timings between last Begin and preceding End.
-	end := handler.s[len(handler.s)-8].(*stats.End)
-	begin := handler.s[len(handler.s)-7].(*stats.Begin)
+	end := handler.s[8].(*stats.End)
+	begin := handler.s[9].(*stats.Begin)
 	diff := begin.BeginTime.Sub(end.EndTime)
 	if diff < 10*time.Millisecond || diff > 50*time.Millisecond {
 		t.Fatalf("pushback time before final attempt = %v; want ~10ms", diff)

